# Série 2
Ce document contient les différents exercices à réaliser. Veuillez compléter et rendre ces exercices dans deux semaines.

Pour chaque exercice:
* implémentez ce qui est demandé
* commentez votre code
* expliquez **en français** ce que vous avez codé dans la cellule correspondante

Dans vos explications à chacun des exercices, indiquez un pourcentage subjectif d'investissement de chaque membre du groupe. **Des interrogations aléatoires en classe pourront être réalisées pour vérifier votre contribution/compréhension.**

## Exercice 1
Implémentez et testez un algorithme permettant d'inverser une liste d'entiers en utilisant une méthode récursive.

In [14]:
def inverse(elements: list[int]) -> list[int]:
    """
    Returns a new list containing all the elements of the provided list in inverse order.
    """
    if len(elements) < 2:
        return list(elements)

    first, *remaining_elements, last = elements
    return [last, *inverse(remaining_elements), first]


In [15]:
assert inverse([5, 8, 7, 3, 2]) == [2, 3, 7, 8, 5]
assert inverse([8, 5]) == [5, 8]

assert inverse([]) == []
assert inverse([1]) == [1]


### Explications

We define the case in which the list contains less than two elements as the recursion's base case.
The solution for this case is trivial, as the inverse of a list that contains at most one element is the list itself.
Note that we always create a copy of the list, as the function's contract (as we've defined it) requires a new list to be returned.

If the provided list contains two or more elements, we can determine the inverse order recursively.
In every iteration, we determine the list's first and last elements, and return a new list in which the first and last element were swapped.
The order of the remainder (i.e, middle part) of the list is inversed recursively.

We can access a list's first, last, and remaining elements in a single assignment statement ([docs](https://docs.python.org/3/reference/simple_stmts.html#assignment-statements)).
The "starred target" on the left-hand side of the assignment, as it's formally called, is assigned a list containing all elements that were not assigned to the other targets.
If the object on the right-hand side only contains two elements, this list is empty.
We use those three components to construct a new list.
The star operator used in the list literal allows us to spread the list contents into the newly created list (otherwise, we would have a nested list).

## Exercice 2
Implémentez une méthode récursive qui trouve et retourne le plus petit élément d'une liste, où la liste et sa taille sont données en paramètre.

In [16]:
def minEl(m, s: int):
    """
    Returns the minimum element in the list m of size s using recursion.
    """
    # Base case 
    if s == 1:
        return m[0]
    
    # Recursive call to find the minimum element in the list
    minValue = minEl(m, s - 1)

    # check if the current element is less than the minimum element found so far
    if m[s - 1] < minValue:
        # return the current element at index s - 1
        return m[s - 1]
    else:
        # return the minimum element found so far in the recursion 
        return minValue

In [17]:
assert minEl([6,5,3,9,1], 5) == 1
assert minEl([6,5,3,-9,1], 5) == -9
assert not minEl([6,5,3,-9,1], 5) == 1

### Explications

The function `minEl` is used to find the minimum element in a list `m` of size `s` using recursion. It takes two parameters: `m`, which is a list of elements, and `s`, which is an integer representing the size of the list.

The base case of the function is when the size `s` is equal to 1. In this case, there is only one element in the list, so it is by default the minimum. The function directly returns this element (`m[0]`). If `s` is greater than 1, the function calls itself again, reducing the size `s` by 1 (`minEl(m, s - 1)`). This allows the function to handle smaller parts of the list step by step.

After getting the minimum of the first `s-1` elements, the function then compares this minimum (`minValue`) with the current element, `m[s - 1]`, which is the last element of the current part of the list. If `m[s - 1]` is smaller than `minValue`, the function returns `m[s - 1]`. Otherwise, it returns `minValue`.

In summary, the function divides the list into smaller parts until it reaches a size of 1 (the base case). Then, as it moves back up, it keeps comparing to find and return the smallest element. 

#### Example

`m = [3, 1, 4]`

1. **First Call:** `s = 3`  
   The function calls itself with `s = 2`.

2. **Second Call:** `s = 2`  
   The function calls itself again with `s = 1`.

3. **Third Call (Base Case):** `s = 1`  
   The function returns `m[0]`, which is `3`.

4. **Going Back Up (Comparing Elements):**
   - In the second call (`s = 2`), it compares `m[1]` (which is `1`) with `minValue` (which is `3` from the previous call). Since `1` is smaller, the function returns `1`.
   - In the first call (`s = 3`), it then compares `m[2]` (which is `4`) with `minValue` (which is `1`). Since `1` is still smaller, the function returns `1`.

So, the minimum value in the list `[3, 1, 4]` is `1`.


## Exercice 3
Implémentez une méthode récursive qui cherche un élément dans une liste triée en utilisant la recherche binaire. La liste, la taille et l'élément cible sont donnés en paramètre.

In [18]:
def find_item(elements: list[int], target_element: int) -> int | None:
    elements_size = len(elements)

    if elements_size == 0:
        return None

    if elements_size == 1:
        if elements[0] == target_element:
            return 0

        return None

    pivot_idx = elements_size // 2
    pivot_element = elements[pivot_idx]

    if target_element < pivot_element:
        return find_item(elements[0:pivot_idx], target_element)

    right_half_result = find_item(elements[pivot_idx:elements_size], target_element)
    if right_half_result is not None:
        return right_half_result + pivot_idx

    return None

In [19]:
assert find_item([1, 2, 4, 7, 8], 7) == 3
assert find_item([-12, -9, 10, 44, 85, 91], -9) == 1
assert find_item([2, 4, 7, 9], 2) == 0

assert find_item([], 1) is None

assert find_item([1, 2, 4, 7, 8], 9) is None
assert find_item([1, 2, 4, 7, 8], 0) is None
assert find_item([1, 2, 4, 7, 8], 3) is None


### Explications

The first conditional statement deals with the edge case in which the provided list is empty.
Neither the regular nor the base case of the recursion can deal with this.

The recursive base case is given by the second conditional statement.
It deals with the situation in which the provided list only contains a single element.
Determining whether the list contains the target element is then as simple as comparing the single element with the target element.
The target element's index hence always is zero.

The task of the recursion's regular case therefore is to split the list into smaller list until the base case is reached.

In every iteration, we compare the list's center-most element (so-called pivot element) with the target element.
If the target element is smaller than the pivot element, we know that the target element must have a lower index than the pivot element.
That is, we can discard all elements with an index equal to or larger than the pivot element's index (called "left half" hereafter).
Likewise, we can discard all elements with an index smaller than the pivot element's index (called "right half" hereafter) in case the target index is equal to or larger than the pivot element.

If we discard the left half, we need to add the size of the left half to the recursive call's result (unless it is `None`).
This accounts for the sub-list's position in the overall list.
Otherwise, we would always obtain an overall result of `0` (or `None` ).

_Note: The parameter that indicates the size of the list has been removed. It did not prove useful for the chosen implementation, and was a potential source of error._

## Exercice 4
La "Fonction 91 de McCarthy" est définie comme suit:

    M(n) for integers > 0:
      if n > 100, M(n) = n - 10
      if n <= 100, M(n) = M(M(n+11))

La notation `M(M(n+11))` est un appel récursif imbriqué.

Implémentez et testez une méthode python qui retourne la nombre de McCarthy.

In [20]:
def mcCarthy(n: int):
    """
    Returns the result of the McCarthy 91 function for the given input n.
    """
    if n > 100:
        # if n > 100, M(n) = n - 10
        return n - 10

    # if n <= 100, M(n) = M(M(n + 11))
    return mcCarthy(mcCarthy(n + 11))

In [21]:
assert mcCarthy(91) == 91
assert mcCarthy(101) == 91
assert mcCarthy(102) == 92
assert mcCarthy(104) == 94

### Explications

The function `mccarthy` is used to calculate the McCarthy value for a given integer `n`. The function is defined recursively, following the rules of the McCarthy function.

The base case of the function is when `n` is greater than 100. In this case, the function returns `n - 10`. If `n` is less than or equal to 100, the function calls itself recursively with `n + 11` as the argument. This is the nested recursive call `mccarthy(mccarthy(n + 11))`.

*More details at Exercice 4.1*

### Exercise 4.1
Quels sont les nombres de McCarthy pour: 1, 15, 79, 99, 100, 101, 200 ?

In [22]:
numbers = [1, 15, 79, 99, 100, 101, 200]

for n in numbers:
    print(f"M({n}) = {mcCarthy(n)}")

M(1) = 91
M(15) = 91
M(79) = 91
M(99) = 91
M(100) = 91
M(101) = 91
M(200) = 190


In [23]:
numbers = [102, 103, 110, 150, 200]

for n in numbers:
    print(f"M({n}) = {mcCarthy(n)}")

M(102) = 92
M(103) = 93
M(110) = 100
M(150) = 140
M(200) = 190


We can see that for any `n` smaller than 100, the McCarthy function will return 91. For `n` greater than 100, the function will return `n - 10`. (So if `n` is 101, the function will return 91 as well but if `n` is 200, the function will return 190.)

#### Why it "converges" to 91
When `n < 100`, the function calls itself in a nested way, meaning it calls `M(n + 11)`, and then calls the function again on that result. This continues until the value becomes greater than **100**, at which point the function starts to follow the first rule and subtracts **10**.


##### Example
For example : `n=99`
1. M(99) = M(M(99 + 11)) = M(110)
2. To calculate M(110), since 110 > 100, we have: M(110) = 110 - 10 = 100 
3. So, M(99) = M(100).

Now, for \( M(100) \):

4. M(100) = M(M(100 + 11)) = M(111)
5. Since 111 > 100, M(111) = 111 - 10 = 101.
6. So, M(100) = M(101).

Similarly:

7. M(101) returns 101 - 10 = 91

So, M(100) = M(101) = 91


#### Conclusion
For any value n < 100, the McCarthy 91 function always returns 91 due to its recursive structure. Each recursive call adds 11 until the value exceeds 100, after which the function subtracts 10 until it reaches 91. This creates a "trap" that ensures all inputs n < 100 converge to 91. For n > 100, the function simply returns n - 10.

[Source](https://en.wikipedia.org/wiki/McCarthy_91_function)

## Exercice 5 - (<font color='#db60cf'>Bonus</font>) Complexité algorithmique de Fibonacci récursif
On définit l'*algorithme de Fibonacci récursif* comme suit :

In [24]:
def fibonacci_recursive(n):
    if n <= 1:
        return n
    else:
        result1 = fibonacci_recursive(n-1,)
        result2 = fibonacci_recursive(n-2,)
        return result1 + result2  # Summation counts for 1 in the complexity

Soit $T(n)$ la complexité temporelle de la fonction Fibonacci lorsqu'elle est appelée avec un $n\in \mathbb{N}$. Il vient de l'implémentation que $T(0)=T(1)=1$ et $$T(n)=T(n-1)+T(n-2)+1$$ pour tout $n>1$.

En admettant que pour tout $n>2$, $T(n-2) \approx T(n-1)$ (en réalité, $T(n-2)=O(T(n-1)$), exprimer $T(n)$ en fonction de $T(n-1)$.
Exprimer alors $T(n-1)$ en fonction de $T(n-2)$, puis $T(n-2)$ en fonction de $T(n-3)$. Enfin, exprimer $T(n)$ en fonction de $T(n-3)$.

Démontrer par récurrence une expression de $T(n)$ en fonction de $n$. L'implémenter dans la cellule contenant `def T(n)`.

<< A REMPLIR PAR L'ETUDIANT >>

In [25]:
# Rendering latex equations with a package installed by magic cell
%pip install latexify-py==0.2.0

ERROR: Ignored the following versions that require a different python version: 0.0.4 Requires-Python >=3.6, <3.9; 0.0.5 Requires-Python >=3.6, <3.9; 0.0.6 Requires-Python >=3.6, <3.9; 0.0.7 Requires-Python >=3.6, <3.9; 0.1.0 Requires-Python <3.11,>=3.7; 0.1.1 Requires-Python <3.11,>=3.7; 0.2.0 Requires-Python <3.11,>=3.7; 0.2.0b1 Requires-Python <3.11,>=3.7; 0.2.0b2 Requires-Python <3.11,>=3.7; 0.3.0b1 Requires-Python <3.12,>=3.7; 0.3.1 Requires-Python <3.12,>=3.7; 0.4.0 Requires-Python <3.12,>=3.7; 0.4.1 Requires-Python <3.12,>=3.7; 0.4.2 Requires-Python <3.12,>=3.7; 0.4.3.post1 Requires-Python <3.12,>=3.7
ERROR: Could not find a version that satisfies the requirement latexify-py==0.2.0 (from versions: none)
ERROR: No matching distribution found for latexify-py==0.2.0
Note: you may need to restart the kernel to use updated packages.


In [26]:
import latexify

@latexify.function
def T(n):
    # A CORRIGER
    return n

T

ModuleNotFoundError: No module named 'latexify'

Ajouter une ligne à `fibonacci_recursive` pour que le troisième subplot (voir ci-après) montre les valeurs appropriées :

In [ ]:
def fibonacci_recursive(n):
    if n <= 1:
        return n
    else:
        result1 = fibonacci_recursive(n-1,)
        result2 = fibonacci_recursive(n-2,)
        # A COMPLETER
        return result1 + result2

### Explications

<< A REMPLIR PAR L'ETUDIANT >>

In [ ]:
import time
import matplotlib.pyplot as plt
import numpy as np


def measure_execution_time(func, *args):
    start_time = time.time()
    result = func(*args)
    end_time = time.time()
    execution_time = end_time - start_time
    return result, execution_time

# Generate values for n
n_values = np.arange(1, 30)

# Lists to store results
execution_times = []
time_complexity_values = []
recursive_calls = []

# Measure execution time, compute time complexity, and count recursive calls for each n
for n in n_values:
    call_counts = []
    _, execution_time = measure_execution_time(fibonacci_recursive, n)
    time_complexity = T(n)
    execution_times.append(execution_time)
    time_complexity_values.append(time_complexity)
    recursive_calls.append(np.sum(call_counts))

plt.figure(figsize=(15, 5))

plt.subplot(1, 3, 1)
plt.plot(n_values, execution_times, label='Execution Time')
plt.xlabel('Input Size (n)')
plt.ylabel('Execution Time (s)')
plt.title('Algorithm Execution Time')

plt.subplot(1, 3, 2)
plt.plot(n_values, time_complexity_values, label=r'$' + T._latex + '$', color='red', linestyle='dashed')
plt.xlabel('Input Size (n)')
plt.ylabel('Time Complexity')
plt.legend()
plt.title('Time Complexity Function')

plt.subplot(1, 3, 3)
plt.plot(n_values, recursive_calls, label='Recursive Calls')
plt.xlabel('Input Size (n)')
plt.ylabel('Number of Recursive Calls')
plt.title('Number of Recursive Calls')

plt.tight_layout()
plt.show()

Que montrent ces courbes ?

<< A REMPLIR PAR L'ETUDIANT >>